### PR TITLE
Jiterator cache for Windows

### LINKED ISSUE #71967: Windows: Support Jiterator Cache
### 🚀 The feature, motivation and pitch

Jiterator Cache is a useful feature that is currently restricted to POSIX compliant systems.
The only reason is that the POSIX APIs access, mkdir and getpid are called.

I understand there is a TODO for c++17 support (for std::filesystem). However, as a temporary measure, we can simply use _access, _mkdir and _getpid for Windows.

### Alternatives

```C
#ifdef _WIN32
  #include <process.h>
  #include <direct.h>
  #include <io.h>
  #define access _access
  #define getpid _getpid
  #define R_OK    4
  #define W_OK    2
  #define F_OK    0
#else
  #include <sys/types.h>
  #include <sys/stat.h> // mkdir
  #include <unistd.h>
#endif
```

```C
#ifdef _WIN32
    if (_mkdir(p_cache_dir) != 0) {
#else
    if (mkdir(p_cache_dir, S_IRWXU | S_IRWXG | S_IRWXO) != 0) {
#endif
```

A suitable cache directory also has to be selected, rather than $HOME that is used currently. In my workaround, I am using $TEMP.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by xsacha**
Fixes #71967

### CODE REVIEW

File: aten/src/ATen/native/cuda/jit_utils.cpp [hunk 1]
```diff
@@ -811,14 +818,18 @@ c10::optional<std::string> get_cache_dir() {
     return {};
   }
 
-  // Cache path comes from PYTORCH_KERNEL_CACHE_PATH, then XDG_CACHE_HOME, then HOME environment variables
+  // Cache path comes from PYTORCH_KERNEL_CACHE_PATH, then TEMP or XDG_CACHE_HOME, then HOME environment variables
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/jit_utils.cpp:821]: can you please mention in this comment that `TEMP` is used only on windows (and `XDG_CACHE_HOME` on linux systems), and make a similar comment in the cuda note that describes caching behavior?

File: aten/src/ATen/native/cuda/jit_utils.cpp [hunk 2]
```diff
@@ -18,13 +18,21 @@
...
 
-#if BUILD_JITERATOR_WITH_CACHE
-  // Uses POSIX headers, which is why these are guarded behind BUILD_JITERATOR_WITH_CACHE
-  // TODO: C++17 has the fileystem header, which may replace these
+// TODO: C++17 has the fileystem header, which may replace these
+#ifdef _WIN32
+  #include <process.h>
+  #include <direct.h>
+  #include <io.h>
+  #define access _access
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/jit_utils.cpp:26]: can you add a comment here on why you are doing this defines? It's a clever trick to reuse existing code, but it'll be nice to describe it explicitly.

**Review by ngimel — COMMENTED** [review]
Thank you, this is great! @mruberry and I looked at it and left a few minor comments.

File: aten/src/ATen/native/cuda/jit_utils.cpp (hunk 2 above)
Comment (xsacha) [reply aten/src/ATen/native/cuda/jit_utils.cpp:26]: Thanks. It's a fairly common code pattern (also used in third_party/tbb/examples/pipeline/square/gen_input.cpp)
Just added an explainer on top.

**Review by mruberry — APPROVED** [review]
Awesome! Thanks @xsacha!

**Comment by anjali411** [comment]
@mruberry  could you please add the release note labels to this PR?

**Comment by mruberry** [comment]
> @mruberry could you please add the release note labels to this PR?

This PR should be rolled up with other Jiterator PRs, but release note and topic labels are added just in case

**Pull request closed without merging**